src/gridfs/download.rs

```diff
@@ -139,16 +140,196 @@ impl GridFsBucket {
+
+        let future = match &mut stream.state {
+            State::Idle(ref mut cursor) => {
+                let new_future = get_bytes(
+                    cursor.take().unwrap(),
+                    buf.len(),
+                    stream.file.length,
+                    stream.file.chunk_size,
+                )
+                .boxed();
+                stream.state.set_busy(new_future)
+            }
+            State::Busy(future) => future,
+            State::Done => return Poll::Ready(Ok(0)),
+        };
+
+        let (result, cursor) = match future.poll_unpin(cx) {
+            Poll::Ready(result) => result,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        stream.state = if !cursor.is_exhausted() {
+            State::Idle(Box::new(Some(cursor)))
+        } else {
+            State::Done
+        };
+
+        match result {
+            Ok(bytes) => {
+                buf[..bytes.len()].copy_from_slice(&bytes);
+                Poll::Ready(Ok(bytes.len()))
+            }
+            Err(error) => Poll::Ready(Err(error.into())),
+        }
+    }
+}
+
+async fn get_bytes(
+    mut cursor: Cursor<Chunk<'static>>,
+    buf_size: usize,
+    expected_file_len: u64,
+    chunk_size: u32,
+) -> (Result<Vec<u8>>, Cursor<Chunk<'static>>) {
+    let bytes_result = get_bytes_inner(&mut cursor, buf_size, expected_file_len, chunk_size).await;
+    (bytes_result, cursor)
+}
+
+async fn get_bytes_inner(
+    cursor: &mut Cursor<Chunk<'static>>,
+    mut buf_size: usize,
+    expected_file_len: u64,
+    chunk_size: u32,
+) -> Result<Vec<u8>> {
+    let mut bytes = vec![];
+    let mut current_n = 0u32;
+    while cursor.advance().await? {
+        let chunk = cursor.deserialize_current()?;
+
+        if chunk.n != current_n {
+            return Err(ErrorKind::GridFs(GridFsErrorKind::MissingChunk { n: current_n }).into());
+        }
+
+        let expected_len = FilesCollectionDocument::expected_chunk_length_from_vals(
+            expected_file_len,
+            chunk_size,
+            current_n,
+        );
+        if chunk.data.bytes.len() != (expected_len as usize) {
+            return Err(ErrorKind::GridFs(GridFsErrorKind::WrongSizeChunk {
+                actual_size: chunk.data.bytes.len(),
+                expected_size: expected_len,
+            })
+            .into());
+        }
+
+        let bytes_to_write = std::cmp::min(chunk.data.bytes.len(), buf_size) as usize;
+        bytes.extend_from_slice(&chunk.data.bytes[..bytes_to_write]);
+
+        current_n += 1;
+
+        buf_size -= bytes_to_write;
+        if buf_size == 0 {
+            break;
+        }
+    }
+
+    if buf_size > 0
+        && (current_n as u64) < FilesCollectionDocument::n_from_vals(expected_file_len, chunk_size)
+    {
+        return Err(ErrorKind::GridFs(GridFsErrorKind::MissingChunk { n: current_n }).into());
+    }
+
+    Ok(bytes)
+}
+
+impl From<Error> for futures_util::io::Error {
```

Comment:
Is the only usage of this impl on line 224? If so (or if there are only a few usages), I think it might be better to keep this as an internal function rather than a public trait impl to avoid bleeding into our public API.